pkg/reconcile/reconcilers_test.go

```diff
@@ -307,9 +316,90 @@ func TestPodSecurityContext(t *testing.T) {
 	assert.True(t, same, "podSecurityContext expected at pod level")
 }
 
-func newReaperWithCassandraBackend() *api.Reaper {
-	namespace := "service-test"
-	reaperName := "test-reaper"
+func TestReaperDeploymentStrategy(t *testing.T) {
+	// Test to ensure that if we set DeploymentStrategy to Rolling then it gets set back to Recreate by reconcile()
+	// Reaper test resource
+	testReaper := newReaperWithCassandraBackend("default", "test-reaper")
+	runAsUser := int64(8675309)
+	testReaper.Spec.Image = "test/reaper:latest"
+	testReaper.Spec.PodSecurityContext = &corev1.PodSecurityContext{
+		RunAsUser: &runAsUser,
+	}
+	// Create CassDC
+	testCassDC := newCassDC("default", "dc1")
+
+	// Test Scheme
+	localSchemeBuilder := runtime.SchemeBuilder{
+		api.AddToScheme,
+		appsv1.AddToScheme,
+		cassdcapi.AddToScheme,
+	}
+	AddToScheme := localSchemeBuilder.AddToScheme
+	testScheme := scheme.Scheme
+	AddToScheme(testScheme)
+
+	// Mock client
+	k8sClient := fake.NewFakeClientWithScheme(testScheme, testCassDC.DeepCopy())
+
+	// Other fixtures
+	testSecretsmngr := NewSecretsManager()
+	testctx := context.Background()
+	testStatusManager := &status.StatusManager{Client: k8sClient}
+	testLogger := mocks.NewTabLogger()
+
+	// Test reconcile deployment
+	r := defaultReconciler{
+		k8sClient,
+		testScheme,
+		testSecretsmngr,
+	}
+	req := ReaperRequest{
+		testReaper,
+		testLogger,
+		testStatusManager,
+	}
+	r.ReconcileDeployment(testctx, req)
+
+	// Ensure Reaper is deploying with RecreateStrategy
```

Comment:
This check should be in the envtests also. It's a single line there to verify the strategy after reaper-operator has created the deployment.

Reply:
If we were to test these changes in the envtests I would counsel doing so via `suite_test.go`. I would also suggest we calibrate the test to a different level of abstraction and target an assertion of the form `assert.Never(t, len(podList.Items) > 1))` (that's pseudocode obviously). 

I suggest such a test be completed as a separate PR so that existing work can be merged ASAP, as it has been pending for a long time. Additionally, I will need to determine how we can express an intention like `assert.Never`, and whether that is even possible.

This PR does already have testing by way of unit tests, which I'd hope would be sufficient coverage to justify a merge.

Reply:
reaper-operator uses gomega + ginkgo at this point. The correct method is called "Consistently":

https://pkg.go.dev/github.com/onsi/gomega#Consistently

These additional unit tests can be in additional PR, but the envtests (controllers/reaper_controller_test.go) should be in this PR. That is the correct method of testing in reaper-operator and that's also a test I would require, since it tests the functionality will correctly work against api-server. These unit tests do not sadly test that.

Reply:
Many thanks for the pointer Michael, `Consistently()` could well be what we need. If we add these tests to `suite_test.go` then I see no issues with using Ginkgo/Gomega given they are already in use there.

Before discussing which PR any integration tests get put in, I think my priority is understanding your comments RE (1) the `retainKeys` merge strategy and (2) whether we need the `testLogger` mock. It sounds like you have thoughts on both fronts that we need to explore before we can make progress?

Reply:
Also, regarding `Consistently()`, how do we call this? Do we need to do so in a goroutine? Otherwise we can't assert that `len(podList.Items) > 1` until after we've already called `Reconcile()` 🤔

Reply:
I mean, integration tests should be in this PR. Unit tests can follow later. The main testing strategy in this project is the integration tests using envtest. And it's the only way to ensure this implementation works (the fakeclient is not something we can trust since it doesn't behave like the correct server).

Neither the goroutine running (but yes, that would be the case with a blocking runner) nor the testLogger are an issue with the envtests. After adding those few test cases to the integration tests, I'm not entirely sure if there's any need for the unit tests.

Reply:
I think it might be due to the time elapsed but I'm a little unclear here too. 

@burmanm can you remind me: where do we want the additional single line specifying that the deployment strategy is `Recreate`? Does this test anything in practice given that there is a unit test already in place?

Reply:
The argument is that the unit test is pointless where as the envtest is more important. This is because this unit test did not catch the fact that the Patch did not even work.

We need a test that verifies it works. Not "it works on my machine after controller-runtime upgrade".

Reply:
I'm not following your point here at all. The controller runtime upgrade is merged into master. It has nothing to do with my machine. The only relevance is that it was the controller runtime upgrade which made `client.StrategicMergeFrom()` available, as that function didn't exist in earlier versions of the module.

The unit test did catch that the patch worked. The difference between a regular patch and a strategic merge patch is that you thought that the regular patch might not be suitable somehow (although I'm still a little vague on this).

In either event, you ended up with a deployment which had a Recreate strategy, so the unit test passes.

I'm asking what you want me to do (beyond the existing test) to verify that this works? The real high level test on this should be the (existing) upgrade tests that live in k8ssandra, since these changes are supposed to fix an issue that was occuring when upgrading the operator.

Reply:
The unit test only catched if the Patch was called, it did not verify if such value can actually be patched using that method or not since it's not calling Kubernetes. Envtests catch how it actually behaves, will it actually patch the object on the api-server. That's what counts. 

Envtest is more important than this unit test, so always start with envtest and keep the unit test if you think it adds value.

So I'm still asking for those envtest changes.